### DISCUSSION TITLE
Documentation discussion(s)

### DISCUSSION THREAD

**Discussion by kcgen** [dosbox-staging/dosbox-staging#1639] [Ideas]
Documentation currently comes in several forms:
 - README (text-format covering general topics and usage, but not translated)
 - Conf file (only up-to-date when the user creates a new one. Benefit is that it's translated)
 - Inline `Z:> intro` help screens, which includes command listing and hotkeys (also translated). 
 - Inline command help via `/?` (useful for learning about custom commands, like `imgmount`, etc. Also translated)
 - Platform build notes and dev-related topics on the wiki.
 - Wiki(s), which include many sources:
     - The [official DOSBox wiki ](https://www.dosbox.com/wiki/), the most thorough coverage of general DOSBox features
     - [The DOSBox-X wiki,](https://dosbox-x.com/wiki/) excellent in-depth coverage and how-to for many common tasks
     - [Staging's wiki](https://github.com/dosbox-staging/dosbox-staging/wiki), which focusing on some of its add-on features. 

Ensuring that new features and changes are reflected in one form of documentation is currently a manual process, and things can fall through the cracks.

Are there better tools or techniques that can improve the above documentation-related maintenance tasks?

**Comment by dreamer** [comment]
There's also a man file (for Unix-like platforms), and changelog is also a kind of documentation.

> So the question is: are there low-effort tools or techniques that can simplify / improve / automate / make-more-reliable, any of the above documentation-related maintenance tasks?

For Rust… yes: https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html… For C++ - not that I know of :(

Anyway; in my last work project we were using changelogs in this format: https://keepachangelog.com/en/1.0.0/ + special program for "bumping version" that automated all the steps required for releasing new version of the software - adding necessary changelog entries was part of the code review, and the content was being updated *during code reviews*, thus release process was really quick. On one side it's "kind-of" step back to having a changelog file in the root project dir… but on the other side - this format is well structured and easy to maintain.

Personally I think we should move in the direction of **replacing**:
- README file
- documentation generated .conf file
- (to a certain extent) wiki
- extensive descriptions in changelog
- several sections in man file

with a new, single documentation source. Not a wiki (because wiki articles tend to grow uncontrollably and it's important for documentation to be brief and readable), but rather something more "book-like" with careful review for every update. I imagine something similar to these systems:

- https://doc.rust-lang.org/book/ (https://github.com/rust-lang/book)
- https://mesonbuild.com/Manual.html
- https://docs.python.org/3/ (in python docs I particularly like the option of switching languages and versions in top-left)

I am talking only about user-facing documentation; information for developers and build-related stuff should stay as it is IMHO.

**Reply by johnnovak** [reply]
I also think that the user-facing manual should *not* be a wiki, but rather a book-like document written in a "conversational" style. Reference style documents are only good for advanced users already familiar with the subject; personally, I find "guide style" documentation best when I'm learning something new. Best of all, you can also generate an index from a user guide (if using the right tool); which would then effectively double as a reference manual.

[The Debian Administrator's Handbook](https://debian-handbook.info/browse/stable/) is a stellar example of this, it's one of the best docos I've ever read. It gives you just enough background information to keep things interesting, but it's never too wordy. It takes you gently through all the subjects it covers without making you feel stressed or overwhelmed. In fact, it makes complex topics appear quite simple, which is a hallmark of great documentation. I really really like it, it's simply a joy to read.

The [Sound](https://www.dosbox.com/wiki/Sound) page on the DOSBox Wiki is an almost perfect example of this style. It's informative, contains some interesting background info for novices unfamiliar with MS-DOS era PCs, and even more experienced users can learn a few new things from it. I think we should just take it and enhance it a bit (I'm assuming no one would object to taking bits and pieces from the various other DOSBox documentation projects, which we would then enhance and add to, of course).

For the tool to use, [Sphinx](https://www.sphinx-doc.org/en/master/) would be my first choice. It's a nice tool specifically aimed at large-scale documentation writing, it supports cross-references, automatically generated index pages, different output formats (e.g. single HTML, multi-page HTML, PDF, LaTeX, etc.) Then there's little things you wouldn't think of when you start off that come really handy later (e.g. markup for keyboard shortcuts). Lots of high profile projects use it, such as [Blender](https://docs.blender.org/manual/en/latest/index.html), [ScummVM](https://docs.scummvm.org/en/v2.5.1/) and virtually all Python projects, naturally.

I'm quite familiar with Sphinx, I've used it quite a bit over the last decade. I've always dabbled with graphic and web design, so I have some experience with creating custom themes for it. Here's a few examples of that:

Documentation of my RPG map editor (not public yet, so handle with care 😛 I think I'll release it finally in a few days...)
https://gridmonger.johnnovak.net/manual/basic-editing.html

Some old doco I wrote for a tool of mine ages ago:
https://twyg.johnnovak.net/docs/install/

Anyway, there are other tools too, and whatever tool or format we choose, I'm happy to contribute!

------ 

A few more related links of interest:

**Google developer documentation style guide**
https://developers.google.com/style

This was suggested reading for every dev where I work. I've read most of it and found it quite useful; it helps you develop a much more consistent and clear style (with practice, of course).

**Hemingway App**
https://hemingwayapp.com/

_"If I had more time, I would have written a shorter setter"_ — I like this saying a lot! It takes effort to develop a succinct and clear writing style, it doesn't come naturally to most people (especially not developers). This app helps with that: you just paste in some text, and it will make ruthless suggestions on stripping everything down to the bare essence. Even if you won't use it all the time (I don't), it's very educational to run a few documents you've written through it and try to make all the suggested corrections. It can be really hard! Again, it was recommended by the technical writer guys at work.

**Comment by kcgen** [comment]
Regarding the new documentation system; I'm not able to progress this (not for lack will, more out of time/health constraints -- and certainly don't want to hold it up!). 

@johnnovak, please feel free to run with it when you feel the time is right.  Your existing readthedocs projects are top notch 👍.   (You've got maintainer rights on the project and organization, so we can setup new repos/gitpages/RTD instances -- hopefully lots of tools at our disposal!). 

For the Staging website (or 'landing page'), currently a GitHub workflow uses Zola to generate static HTML pages from markdown content (on-push events), and some JavaScript to create the latest `main` download links. 

Here's that repo: https://github.com/dosbox-staging/dosbox-staging.github.io

Just sharing this as some background, which we can leverage or reuse as we see fit.

**Comment by johnnovak** [comment]
After much deliberation and conferring with the technical writer at my work, I've come to the conclusion that it would best to use [MkDocs](https://www.mkdocs.org/) for the user-facing documentation. While I personally don't mind Sphinx, the MkDocs project has a lot more traction, and the Markdown format is preferred by most people over reStructuredText. The goal here is to make people _want_ to contribute to the doco, so I'm in favour of anything that gets us closer to that goal!

The [mkdocs-material](https://squidfunk.github.io/mkdocs-material/) theme is used by countless high-profile projects—if it's good enough for them, it will be good for us too. It's extremely high-quality, and it ticks all the important boxes in terms of what we want from our documentation tool. I won't list all the important features here, just read the [front page](https://squidfunk.github.io/mkdocs-material/) first, then have a look at the list of supported features in the [reference](https://squidfunk.github.io/mkdocs-material/reference/
), it's very comprehensive.

The theme has been built with hackability in mind, it's very clearly laid out, and it's meant to be modified and customised. I'm no stranger to hacking CSS, webdesign, and design in general, so we'll be able to shape the default template in any way we want to give it some custom "DOSBox Staging" touch. But even just the stock theme has a lot of customisable options, e.g. play around with the colours [here](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/), switch it to the dark theme, try different primary colours, etc.

So unless objections, I am going to set up a standard MkDocs project with a default Material theme in the coming weeks. Then I'll start shoveling over some of the existing doco from the wiki and other various sources. We can discuss the general structure in detail and who does what as we go. I am fine taking care of the sound section (but help/input from others is always welcome in that area, of course).

**Reply by kcgen** [reply]
Haven't heard of mkDocs - wow, really nice!

I've now read through all the links, and it looks like a no-nonsense, practical /and/ elegant system for writing and managing really nice documentation. 

It looks about as good as it gets!  For some years, I  lamented the huge gulf between the full markup-driven documentation approach (and managing the backend to such a system), versus the old school WYSIWIG `.doc/.pdf` approach.  It looks like MkDocs has minimized a lot of that prior aspect - exactly what we need.

> While I personally don't mind Sphinx, the MkDocs project has a lot more traction, and the Markdown format is preferred by most people over reStructuredText. The goal here is to make people want to contribute to the doco, so I'm in favour of anything that gets us closer to that goal!

Couldn't agree more!  Thanks for kicking this off, @johnnovak !

**Comment by pengowray** [comment]
Just chiming in with another _yes, please do docs_. The official DOSBox wiki is dead. New docs are missing and needed.

The official DOSBox wiki is dead static pages at this point. In 2017, there was one singular game added to the list of supported games, and that's been the only edit to the list since 2013. The releases page says the latest DOSBox release was 2010 (and links to it). The _[Sound](https://www.dosbox.com/wiki/Sound)_ page, mentioned above, is definitely a good informative page and has a nice style, but its most recent edit was in 2015 [exactly 7 years ago today 🎂🥳], and is almost certainly outdated. No pages at all have been edited in the last 100+ days (MediaWiki doesn't show beyond that). It's a wiki so why don't I fix it? The last time a new wiki account was created was 2018 and there's no information on who to give a secret handshake to if you want to join the secret club who can help update the docs. (I even tried asking on Vogons but only could get info on how to join their own wiki, which I should mention is yet another good source)

Recently, I found the official DOSBox wiki's _[serial ports](https://www.dosbox.com/wiki/Configuration:SerialPort)_ page was simply ancient, missing vital info on the the settings I was looking for. So instead I needed to use Staging's source code comments as documentation. (Staging seems to have the most easily accessible source code). Which was great and all (it's very well commented), but really it would be nice to have "real" documentation.

By the way, I've created a new [Serial Ports](https://github.com/dosbox-staging/dosbox-staging/wiki/Serial-Ports) page to dosbox-staging's github wiki. It's no meandering conversational guide, but at least it's some kind of documentation. There are a lot of serial port options that just weren't documented anywhere. Please feel free to hit the edit button.

I'm a fan of wikis—when they're editable—because they let many people work on improving the docs, in a piecemeal fashion, with a low barrier of entry. I've seen Sphinx recommended before, but it seems very intimidating for anyone wanting to dip their toe in by contributing a minor edit or two. Considering not a lot documentation seems to get written very often, a wiki seems like a good way to go.

The problem with github's wikis in particular though is they're largely invisible to Google due to the no-spiders policy. If you're lucky you'll find a page in search results via some third party mirror site. Also the formatting is a little restrictive.

If there are people dedicated to working on the documentation, then there's a lot of good suggested frameworks in comments above (MkDocs sounds as good as any). Just please make sure anyone can make a PR. 

Biggest thing really is to start somewhere, and actually get started.

**Reply by johnnovak** [reply]
I agree with you on pretty much everything you wrote above @pengowray. A documentation project is sorely needed, indeed. You might be delighted to hear that I'm determined to get the MkDocs based doco project off the ground this month. It will probably not look very polished in the beginning, but it will be a start.

>  Just please make sure anyone can make a PR.

Yes that will be possible. Consistency of style and tone definitely needs to be ensured, though (in addition to providing useful and technically accurate information, of course), but that's what the PR process is for. I'm personally interested in fleshing out the sound related sections and establishing some sort of general structure we can then build upon. Creating high-quality documentation is a big undertaking that also requires ongoing maintenance, so we'll be happy for all the help we can get. You're most certainly very welcome to contribute, should you wish to do so — and judging by your enthusiasm, I'm hoping that's the case! 😄 

(By the way, I'm also Hungarian and live in Brisbane, Australia! 😎 🌴 🏖️  )

**Reply by xga-cga** [reply]
Suggestions:
 put documentation in build releases or another repo by version number.  
DOSbox has an old compatability list.  Vogons forum has configuration and troubleshooting info.  OpenLibrary is an available website to host data.

**Comment by johnnovak** [comment]
# Documentation workflow

This is a proposal for our "branching strategy" and general workflow for the doco. As always comments are welcome, etc 😄 

## Goals & assumptions

- We will maintain a **single documentation** (so no separate versions for every single release)
  - Versioned documentation makes a lot of sense for libraries, but for our project users are expected to migrate to the latest release once it's out
  - Question: Windows and macOS users can always easily upgrade, but are Linux packages a problem? The majority of [Linux repos](https://dosbox-staging.github.io/downloads/linux/) are behind our latest version.
- We want the **code and doco to evolve in lockstep**
  - Every functional change must be accompanied by the relevant doco change in the same PR (of course, this assumes the relevant section already exists; we'll be a bit more relaxed about this until we get there 😅)
  - An exception to this is if the author _really-really-truly-with-sugar-on-the-top_ hates writing documentation, in which case he/she needs to work together with someone else for the doco part. This is to ensure high-quality outcomes and not to force anybody to do tasks they hate doing! 😎 
- From the above follows the **doco must live in the same repo as the code**
  - There are ways around this (git submodules, git-subtree), but I like to keep things simple
- Overall, we want the doco process to be **simple & low-friction**
- We don't want to complicate our current simple merge-to-main, release-from-main branching strategy which has been working well for us.
- ...but we can't just wait 3-6 months for the next doco update until the next release is out, neither can we publish documentation for features intended for the next version! 😈 _← this is the devil in the details!_

## Proposed solution

To keep things simple, we'll just maintain two copies of the MkDocs doco sources in our main repo where we keep our code:

```
/site
  /current
    <mkdocs sources>
    ...
  /dev
    <mkdocs sources>
    ...
```

This will be set up so that whenever there's a new commit in `main`, a  GitHub workflow generates both versions of the site, and publishes them to the `gh-pages` branch of the same repo as follows:
  - The `current` version is published to the root, and is available at https://dosbox-staging.github.io (this is the documentation of our current release, and our current website)
  - The `dev` version is published to the `dev` subdir, and is thus available at https://dosbox-staging.github.io/dev (this is the WIP documentation for our upcoming release, and potentially it could contain major changes to the website, although this is less likely)
  
The nice thing about this is that both versions will be always autogenerated without need for manual intervention, and the `dev` version is publicly available so interested users can have a peek at it. A sticky red banner will be displayed at the top of the `dev` site to make sure no one mistakes it for the current version.

Note that because the website and the user manual are generated from the same MkDocs sources, this separation is necessary in case we want to add a blog section to the website in the future, or we simply need to fix something up on the current site between releases.

## Use cases

So let's see how we would work with this in practice!

### Enhancements & new features 

- New feature gets merged into `main` in one or a series of PRs
- Every PR contains documentation updates or additions to `site/dev` _only_
- After the merge, the new `dev` documentation gets auto-published (and the current documentation is unchanged).

### Fix current documentation, or update current website

- Merge PR containing changes to _both_ `site/current` and `site/dev`.
- After the merge, both `current` and `dev` docos get auto-published.

Just to spell it out: changes made to the `site/current` version must _**always**_ be manually "forward-ported" to `site/dev`. However, this shouldn't happen too often once the documentation is fairly mature. If the files are identical in both versions, the "porting" is process is trivial; if they have diverged a lot, manual resolution is needed anyway as automatic merging rarely works well with English text.

### Release new version

Now watch this! 😎🔥 

```bash
rm -rf site/current/*
cp -R site/dev/* site/current
git add site/current
```

![](https://media.giphy.com/media/Vff5Qxz6LLzag/giphy.gif)

That's how we make the `dev` version `current` at the time of the release (before tagging). It seems drastic, but git will actually figure out a reasonable diff from the above (I do this trick all the time, and it does), then we'll just use the commit message "Publish 0.81.0 documentation" and that's it.

Maybe slighlty unorthodox, but I don't really care because it does the job, and does it in a simple way 🤷🏻 If you really want to see the history of the doco, you still have the full untampered commit log of the `site/dev` branch which is our "true" doco history, if that makes sense (we'll _never_ screw around with `site/dev` this way!)

## Further notes

Note that English text is much harder to merge automatically than code, and diffs are often a bit meaningless, especially if reformatting is involved. Therefore, we could gain very little by having actual `current` and `dev` branches in a separate repo (maybe mapping them into the main code repo with git-subtree or something) and attempting to merge the changes from `current` to `dev` automatically; quite often a manual resolution would be needed anyway. However, this would complicate the PR and deploy process a *lot*, therefore I'm quite convinced it would be a major net loss and it's just simply not worth it for our intended use-cases.

Overall, I like simple solutions that just work and cause minimal complication. The benefit of this solution is that we don't need to complicate our workflows and learn new tricks, just need to adhere to a few simple and easy to remember rules.

If someone has better ideas on how to achieve the above goals, I'm all ears and let the better solution win! 🚀 😄

**Reply by shermp** [reply]
Some great ideas there.

Instead of having the duplicate documentation sources `site/dev`/`site/current`, how about having a single set of documentation, and a `documentation` branch that holds the docs for the current version? The `dev` docs then always come from main, and the current docs come from the documentation branch. Whenever the current docs need updating, just backport them into the branch from `main` and have an action redeploy.

**Reply by johnnovak** [reply]
I like this variation; with this approach we could do proper merging, as opposed to my delete+copy trick.

I'm not too familiar with GitHub workflows/actions, but this should all be doable. Like you said, requiring manual deploys for the `documentation` branch might be even a feature (but I'm pretty sure we could auto-deploy on new commits if we wanted to).

And when there's a new release, we can always just merge `main` to `documentation` which in theory should result in zero conflicts because of the backports. We don't actually care about the code in that branch, but that's fine and it will be easy sailing anyway when it comes to merging on releases (guaranteed conflictless merges for anything that is not documentation related).

**Reply by kcgen** [reply]
Tags are named aliases into the tree at a specific commit. Some tags are meant to be stapled in place, like tag-v0.80.0.

But other tags can be more like sticky notes and are meant to be moved. We could have a tag called "current-docs", that points to the stable release docs, whatever that might be. When we release a new version, we move the tag forward. 

The release process splits off a branch `release/ver-num` from the main line. If you list all the branches in the repo, you will see all these release 0.80, 0.79, etc branches. So the docs/ will now be branched too, in lockstep with the code.

A tag is also applied (v0.80.0) on those release branches so the docs will get these tags too.  Plus, we can now also drop the `current-docs` tag on the latest stable version. 

So when we release 0.80.1 (soon enough!), Current-docs would be stepped forward to it.

**Reply by johnnovak** [reply]
Ok, so you're basically saying `current-docs` would always point to the last released release branch, then if we want to make changes to the stable doco or the website before the next release, we just push stuff into that release branch, then also move `current-docs` forward to it points to the last commint on that branch. That could work.

And the `dev` version would be simply always generated from `main`, like I explained above.
